src/index.js

```diff
@@ -94,38 +94,72 @@ const uploadImage = async (req, res) => {
     return;
   }
 
-  // Valid token
-  const form = new formidable.IncomingForm();
-  const files = await promiseUpload(form, req);
-  if (!files.image || !files.image.path) {
-    res.status(400).end();
-    metric.addTag('status', 400);
-    metrics.write(metric);
-    return;
-  }
+  let uploadCompleted = false;
+  let uploadCancelled = false;
+  // When an uploads gets cancelled in the progress, we want to make the token available again
+  const handleCancelledUpload = async () => {
+    if (!uploadCompleted && !uploadCancelled) {
+      log('info', `Freeing token for ${name} as the upload was aborted prior to completion`);
+      await token.deleteTokenForImageId(name);
+      uploadCancelled = true;
+    }
+  };
+
+  try {
+    req.once('close', handleCancelledUpload);
+
+    // Valid token
+    const form = new formidable.IncomingForm();
+    const files = await promiseUpload(form, req);
+    if (!files.image || !files.image.path) {
+      res.status(400).end();
+      metric.addTag('status', 400);
+      metrics.write(metric);
+      return;
+    }
+
+    const isAllowedToHandle = await imageResponse.hasAllowableImageSize(files.image.path, MAX_IMAGE_IN_MP);
+    if (!isAllowedToHandle) {
+      log('warn', `Image ${name} was too big to handle (over ${MAX_IMAGE_IN_MP} Megapixel) and hence rejected`);
+      res.status(413).end();
+      metric.addTag('status', 413);
+      metrics.write(metric);
+      return;
+    }
+
+    const cropParameters = cropParametersOnUpload(req);
+    if (uploadCancelled) {
+      // If the image is cancelled at this point, save ourselves the trouble of doing any image work
+      await handleCancelledUpload();
```

Comment:
It's not clear why this await is done, since it's only called when `uploadCancelled` is true, while inside `handleCancelledUpload`, `uploadCancelled` needs to be false to do anything? Don't you just need to return here?

Reply:
I definitely do!